### PR TITLE
new module: Configure LVM and NFS file systems for AIX

### PR DESCRIPTION
**Pull request by kairoaraujo**
##### SUMMARY
This module creates, removes, mount and unmount LVM and NFS file system for AIX using ```/etc/filesystems```. For LVM file systems is possible to resize a file system.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
aix_filesystem

##### ANSIBLE VERSION
```
ansible 2.5.0 (aix_filesystem e984ea3edb) last updated 2017/09/24 10:57:14 (GMT +200)
  config file = None
  configured module search path = ['/Users/kairoaraujo/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/kairoaraujo/Dev/Ansible/modules/aix_filesystem/ansible/lib/ansible
  executable location = /Users/kairoaraujo/Dev/Ansible/modules/aix_filesystem/ansible/bin/ansible
  python version = 3.6.2 (v3.6.2:5fd33b5926, Jul 16 2017, 20:11:06) [GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]
```

##### ADDITIONAL INFORMATION
Functional tests.

aix_fs.yml
```yml
---
- name: testing aix_filesystem module
  hosts: 172.16.100.24
  tasks:
    - name: Umounting /testfs
      aix_filesystem:
        filesystem: /testfs
        state: unmounted

    - name: Removing /testfs
      aix_filesystem:
        filesystem: /testfs
        state: absent

    - name: Creating a new file system
      aix_filesystem:
        filesystem: /newfs
        size: 1G
        state: present
        vg: datavg

   - name: Creating NFS filesystem from nfshost (Linux NFS server)
    aix_filesystem:
      device: /home/ftp
      nfs_server: nfshost
      filesystem: /nfs/ftp
      state: present

    - name: Creating a logical volume testlv (aix_lvol module)
      aix_lvol:
        vg: datavg
        lv: testlv
        size: 2G
        state: present

    - name: Create filesystem in a previously defined logical volume
      aix_filesystem:
        device: testlv
        filesystem: /testfs
        state: present

    - name: Create an already existing filesystem using existing logical volume.
      aix_filesystem:
        vg: datavg
        device: mksysblv
        filesystem: /mksysb
        state: present

    - name: Create a filesystem in a non-existing VG
      aix_filesystem:
        vg: nonexistvg
        filesystem: /newlv
        state: present

    - name: Resizing /mksysb to 1G
      aix_filesystem:
        filesystem: /mksysb
        size: 1G
        state: present

    - name: Resizing /mksysb to +512M
      aix_filesystem:
        filesystem: /mksysb
        size: +512M
        state: present

    - name: Resizing /mksysb to 11G
      aix_filesystem:
        filesystem: /mksysb
        size: 11G
        state: present

    - name: Resizing /mksysb to 11G (already done)
      aix_filesystem:
        filesystem: /mksysb
        size: 11G
        state: present

    - name: Resizing /mksysb to -2G
      aix_filesystem:
        filesystem: /mksysb
        size: -2G
        state: present

    - name: Resizing /mksysb to 100G (no enought space)
      aix_filesystem:
        filesystem: /mksysb
        size: +100G
        state: present

    - name: Unmount filesystem /home/ftp
      aix_filesystem:
        filesystem: /home/ftp
        state: unmounted

    - name: Remove NFS filesystem /home/ftp
      aix_filesystem:
        filesystem: /home/ftp
        rm_mount_point: yes
        state: absent

    - name: Mount filesystem /newfs
      aix_filesystem:
        filesystem: /newfs
        state: mounted

    - name: Remove mounted /newfs
      aix_filesystem:
        filesystem: /newfs
        rm_mount_point: yes
        state: absent

    - name: Umount /newfs
      aix_filesystem:
        filesystem: /newfs
        state: unmounted

    - name: Remove /newfs
      aix_filesystem:
        filesystem: /newfs
        rm_mount_point: yes
        state: absent
```
output 

```
ansible-playbook -i ../../inventory ../aix_fs.yml -v
No config file found; using defaults

PLAY [testing aix_filesystem module] **************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************
ok: [172.16.100.24]

TASK [Umounting /testfs] **************************************************************************************************************
ok: [172.16.100.24] => {"changed": false, "failed": false, "msg": "File system /testfs already unmounted.", "state": "unmounted"}

TASK [Removing /testfs] ***************************************************************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "rmlv: Logical volume testlv is removed.\n", "state": "absent"}

TASK [Creating a new file system] *****************************************************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "File system created successfully.\n1048340 kilobytes total disk space.\nNew File System size is 2097152\n", "state": "present"}

TASK [Creating NFS filesystem from nfshost (Linux NFS server)] ************************************************************************
ok: [172.16.100.24] => {"changed": false, "failed": false, "msg": "File system /nfs/ftp already exists.", "state": "present"}

TASK [Creating a logical volume testlv (aix_lvol module)] *****************************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "Logical volume testlv created."}

TASK [Create filesystem in a previously defined logical volume] ***********************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "File system created successfully.\n2096884 kilobytes total disk space.\nNew File System size is 4194304\n", "state": "present"}

TASK [Create an already existing filesystem using existing logical volume.] ***********************************************************
ok: [172.16.100.24] => {"changed": false, "failed": false, "msg": "File system /mksysb already exists.", "state": "present"}

TASK [Create a filesystem in a non-existing VG] ***************************************************************************************
ok: [172.16.100.24] => {"changed": false, "failed": false, "msg": "Volume group nonexistvg does not exist.", "state": "present"}

TASK [Resizing /mksysb to 1G] *********************************************************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "Filesystem size changed to 2097152\n", "state": "present"}

TASK [Resizing /mksysb to +512M] ******************************************************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "Filesystem size changed to 3145728\n", "state": "present"}

TASK [Resizing /mksysb to 11G] ********************************************************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "Filesystem size changed to 23068672\n", "state": "present"}

TASK [Resizing /mksysb to 11G (already done)] *****************************************************************************************
ok: [172.16.100.24] => {"changed": false, "failed": false, "msg": "The filesystem size is already 23068672.\n", "state": "present"}

TASK [Resizing /mksysb to -2G] ********************************************************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "Filesystem size changed to 18874368\n", "state": "present"}

TASK [Resizing /mksysb to 100G (no enought space)] ************************************************************************************
ok: [172.16.100.24] => {"changed": false, "failed": false, "msg": "0516-787 extendlv: Maximum allocation for logical volume mksysblv\n\tis 512.\n", "state": "present"}

TASK [Unmount filesystem /home/ftp] ***************************************************************************************************
ok: [172.16.100.24] => {"changed": false, "failed": false, "msg": "File system /home/ftp already unmounted.", "state": "unmounted"}

TASK [Remove NFS filesystem /home/ftp] ************************************************************************************************
ok: [172.16.100.24] => {"changed": false, "failed": false, "msg": "File system /home/ftp does not exist.", "state": "absent"}

TASK [Mount filesystem /newfs] ********************************************************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "File system /newfs mounted.", "state": "mounted"}

TASK [Remove mounted /newfs] **********************************************************************************************************
ok: [172.16.100.24] => {"changed": false, "failed": false, "msg": "File system /newfs mounted.", "state": "absent"}

TASK [Umount /newfs] ******************************************************************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "File system /newfs unmounted.", "state": "unmounted"}

TASK [Remove /newfs] ******************************************************************************************************************
changed: [172.16.100.24] => {"changed": true, "failed": false, "msg": "rmlv: Logical volume fslv00 is removed.\n", "state": "absent"}

PLAY RECAP ****************************************************************************************************************************
172.16.100.24              : ok=21   changed=11   unreachable=0    failed=0
```

### CODE REVIEW

File: lib/ansible/modules/system/aix_filesystem.py
Comment (jtyr) [inline lib/ansible/modules/system/aix_filesystem.py]: If you want to have this as a new paragraph, you should add `-` at the beginning. The same for the formatting of the description of the `size` option bellow.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (jtyr) [inline lib/ansible/modules/system/aix_filesystem.py:117]: I would separate individual examples by an empty line.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (jtyr) [inline lib/ansible/modules/system/aix_filesystem.py]: You can remove `required=False` as that's the default.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (jtyr) [inline lib/ansible/modules/system/aix_filesystem.py:466]: Please test if `default=False` is the default value when using type `bool`.

**Review by jtyr — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: This is formatted like this:
```python
# Copyright: (c) 2017, Kairo Araujo <kairo@kairo.eti.br>
```

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: This should be on a single line.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: No blank line here.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: `author` is a list, so it should be:
```yaml
author:
- Kairo Araujo (@kairoaraujo)
```

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: The `short_description` must not have a trailing dot. Keep the index clean.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: This should be:
```yaml
    type: bool
```

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: If it's not required, leave this out. It's implied that parameters are not required.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Same here. No need to state this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Again, this should become:
```yaml
type: bool
```

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Get rid of this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: The default value is None, so you can leave this out. Adds no value.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Remove this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Remove this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Remove this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Remove this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Remove this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Remove this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Remove this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: You probably want this here:
```yaml
    type: bool
    default: 'no'
```

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Remove this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py:216]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py:371]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py:428]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [reply lib/ansible/modules/system/aix_filesystem.py:466]: No, the default is None. So you can tell if the parameter was set or not.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: You don't need to state `default=None` as this is the default value.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: You don't need to state default=None as this is the default value.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: You don't need to state default=None as this is the default value.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: You don't need to state default=None as this is the default value.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Add trailing comma.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Add trailing comma.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please put this on a single line. Lines can be 159 characters wide.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: This is useless unless you also return it as a value.

**Review by dagwieers — CHANGES_REQUESTED** [review]
So looking at the code, I think the main part could use a generic **result** dictionary that is modified as you go.

```python
result = dict(
    changed=False
)
```
And then you can set it to changed whenever there's a change induced.
```python
result['changed'] = True
result['msg'] = 'Foo bar'
```
And when you call `fail_json()` or `exit_json()` you can simply provide it the dictionary:
```python
if something:
    fail_json(**result)

exit_json(**result)
```

This ensures that if you make a change, but it fails subsequently, you return the change-flag on failure as well.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (jtyr) [reply lib/ansible/modules/system/aix_filesystem.py]: String or list are allowed here so no change is required.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (kairoaraujo) [reply lib/ansible/modules/system/aix_filesystem.py]: I thought it was optional, anyway. Indeed. :)

Comment (dagwieers) [reply lib/ansible/modules/system/aix_filesystem.py]: Right but this seems arbitrary, cut off at 40 characters ?
For readability I think it's better to not split lines if it is not required.

But you are right, it's not mandatory. So you can ignore this :-)

**Comment by dagwieers** [comment]
It would probably be best to also include integration tests. Not that they will be run as part of Shippable (as we don't have AIX targets), but it is useful for contributors (and maintainers) to test if the module still works as expected.

(And you already have these tests)

**Comment by kairoaraujo** [comment]
Hi @dagwieers 

I'm blocked with integration tests.

Should I write ```ansible-test aix-integration``` to support a specific inventory to run it? (I already tried it but I receive error[1])

Write the integration tests is ok, but how can I run this tests by myself?
Is it possible using ```ansible-test integration --remote ???```

I would like run the test in my environment.

Thanks in advance to help me with that.

[1]
```shell
ansible-test -h
usage: ansible-test [-h] COMMAND ...

positional arguments:
  COMMAND
    integration        posix integration tests
    network-integration
                       network integration tests
    windows-integration
                       windows integration tests
    aix-integration    aix integration tests
    units              unit tests
    compile            compile tests
    sanity             sanity tests
    shell              open an interactive shell
    coverage           code coverage management and reporting

optional arguments:
  -h, --help           show this help message and exit


ansible-test aix-integration --aix test/integration/inventory.aixremote.template aix_filesystem
Traceback (most recent call last):
  File "/Users/kairo/Dev/Ansible/aix_integration/ansible/test/runner/ansible-test", line 696, in <module>
    main()
  File "/Users/kairo/Dev/Ansible/aix_integration/ansible/test/runner/ansible-test", line 92, in main
    args.func(config)
  File "/Users/kairo/Dev/Ansible/aix_integration/ansible/test/runner/lib/executor.py", line 403, in command_aix_integration
    remotes = [instance.wait_for_result() for instance in instances]
  File "/Users/kairo/Dev/Ansible/aix_integration/ansible/test/runner/lib/thread.py", line 48, in wait_for_result
    exec('raise exception[0], exception[1], exception[2]')  # pylint: disable=locally-disabled, exec-used
  File "/Users/kairo/Dev/Ansible/aix_integration/ansible/test/runner/lib/thread.py", line 34, in run
    self._result.put((self.action(), None))
  File "/Users/kairo/Dev/Ansible/aix_integration/ansible/test/runner/lib/executor.py", line 427, in aix_run
    core_ci.start()
  File "/Users/kairo/Dev/Ansible/aix_integration/ansible/test/runner/lib/core_ci.py", line 149, in start
    return self.start_remote()
  File "/Users/kairo/Dev/Ansible/aix_integration/ansible/test/runner/lib/core_ci.py", line 153, in start_remote
    with open(self.ci_key, 'r') as key_fd:
IOError: [Errno 2] No such file or directory: '/Users/kairo/.ansible-core-ci.key'
```

**Review by jtyr — APPROVED** [review]
LGTM

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [reply lib/ansible/modules/system/aix_filesystem.py]: We are moving to the use of lists, because changes to the author list only affects the new line. The same reason we prefer trailing comma's in multi-line lists or dicts in python.

For new modules it's better to start off with lists.

Comment (jtyr) [reply lib/ansible/modules/system/aix_filesystem.py]: 1) There is rarely more than one author.
2) The same principle is used elsewhere in Ansible that you can use either string or list.

**Comment by gforster** [comment]
@mattclay 
@jtyr 
@dagwieers
@jborean93

What needs to be done here to get this merged?

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Please break this up in 2 distinct items. The first is an explanation what the option is, subsequent items clarify how to use it, etc.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py:37]: Since recently we expect module parameters to have the type declared in the documentation.
So this one would state:
```yaml
    type: str
```

**Review by dagwieers — CHANGES_REQUESTED** [review]
LGTM, although there are still a few review items open.
Then we can merge this.

File: lib/ansible/modules/system/aix_filesystem.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_filesystem.py]: Is there any purpose in returning the state that the user already supplied ?
We try to avoid this generally.

**Review by dagwieers — DISMISSED** [review]

**Review by dagwieers — APPROVED** [review]
LGTM, waiting for CI to succeed.

**Pull request merged**
